lib/fast_inserter/fast_inserter_base.rb

```diff
@@ -92,20 +95,44 @@ def fast_insert_group(group)
       end
     end
 
+    def existing_values_sql(group_of_values)
+      sql = "SELECT #{@variable_columns.join(', ')} FROM #{@table_name} WHERE #{values_hash_to_sql(@static_columns)}"
+
+      # Using the database to filter existing values is not enabled by default since large IN queries in the SQL
+      # statement can be bad for database performance.
+      if @options[:filter_existing_in_database]
+        if @variable_columns.length > 1
+          group_of_values_sql = group_of_values.map do |group|
+            values_hash = variable_column_values_to_hash(group)
+            "(#{values_hash_to_sql(values_hash)})"
+          end.join(' OR ')
+
+          sql += " AND (#{group_of_values_sql})"
+        else
+          values_to_check = ActiveRecord::Base.send(:sanitize_sql_array, ['?', group_of_values.flatten])
+          sql += " AND #{@variable_columns.first} IN (#{values_to_check})"
+        end
+      end
+
+      sql
+    end
+
     # Queries for the existing values for a given group of values
     def existing_values(group_of_values)
-      sql = "SELECT #{@variable_columns.join(', ')} FROM #{@table_name} WHERE #{existing_values_static_columns}"
+      sql = existing_values_sql(group_of_values)
 
       # NOTE: There are more elegant ways to get this field out of the resultset, but each database adaptor returns a different type
       # of result from 'execute(sql)'. Potential classes for 'result' is Array (sqlite), Mysql2::Result (mysql2), PG::Result (pg). Each
       # result can be enumerated into a list of arrays (mysql) or list of hashes (sqlite, pg)
       results = ActiveRecord::Base.connection.execute(sql)
       existing_values = stringify_values(results)
 
-      # Rather than a giant IN query in the sql statement (which can be bad for database performance),
-      # do the filtering of relevant values here in a ruby select.
-      group_of_values_strings = stringify_values(group_of_values)
-      existing_values & group_of_values_strings
+      if @options[:filter_existing_in_database]
+        existing_values
+      else
+        group_of_values_strings = stringify_values(group_of_values)
+        existing_values & group_of_values_strings
```

Comment:
@jordondornbos should this only return `existing_values`. i think that currently the logic we end up running on line 88 is `non_existing_values = stringify_values(group) - (existing_values(group) & stringify_values(group)` - it seems like just `non_existing_values = stringify_values(group) - existing_values(group)` is all we need. looks like you already made this change in your new code, so might be worth doing it in this  existing line of code too. up to you just something i noticed during review